sydtest/src/Test/Syd/Diff.hs

```diff
@@ -0,0 +1,525 @@
+
+-- | Diff two vectors with grouped results using a custom equality operator
+getGroupedVectorDiffBy :: forall a b. (a -> b -> Bool) -> Vector a -> Vector b -> Vector (PolyDiff (Vector a) (Vector b))
+getGroupedVectorDiffBy eq old new = computeGroupedDiffFromEditScript old new (getEditScriptBy eq old new)
+
+-- | Compute the edit script to turn a given vector into the second given vector
+getEditScript :: forall a. Eq a => Vector a -> Vector a -> Vector Edit
+getEditScript = getEditScriptBy (==)
+
+-- | Compute the edit script to turn a given vector into the second given vector with a custom equality operator
+--
+-- From https://blog.robertelder.org/diff-algorithm/
+getEditScriptBy :: forall a b. (a -> b -> Bool) -> Vector a -> Vector b -> Vector Edit
+getEditScriptBy eq old new = V.fromList $ DList.toList $ runST $ go old new 0 0
+  where
+    go :: forall s. Vector a -> Vector b -> Int -> Int -> ST s (DList Edit)
+    go e f i j = do
+      -- N,M,L,Z = len(e),len(f),len(e)+len(f),2*min(len(e),len(f))+2
+      let upperN :: Int
+          upperN = V.length e
+
+      let upperM :: Int
+          upperM = V.length f
+
+      let upperL :: Int
+          upperL = upperN + upperM
+
+      let upperZ :: Int
+          upperZ = 2 * min upperN upperM + 2
+
+      -- if N > 0 and M > 0:
+      if upperN > 0 && upperM > 0
+        then do
+          -- w,g,p = N-M,[0]*Z,[0]*Z
+          let w :: Int
+              w = upperN - upperM
+
+          g <- MV.replicate upperZ 0 :: ST s (MVector s Int)
+
+          p <- MV.replicate upperZ 0 :: ST s (MVector s Int)
+
+          -- for h in range(0, (L//2+(L%2!=0))+1):
+          let hs :: [Int]
+              hs = [0 .. ((upperL `quot` 2) + (if odd upperL then 1 else 0))]
+
+          fmap fromJust $ -- We use fromJust because the algorithm is guaranteed to terminate
+            forUntilJust hs $ \h -> do
+              -- for r in range(0, 2):
+              forUntilJust [0, 1 :: Int] $ \r -> do
+                -- c,d,o,m = (g,p,1,1) if r==0 else (p,g,0,-1)
+                let (c, d, o, m) = if r == 0 then (g, p, 1, 1) else (p, g, 0, -1)
+
+                -- for k in range(-(h-2*max(0,h-M)), h-2*max(0,h-N)+1, 2):
+                let lo :: Int
+                    lo = -(h - 2 * max 0 (h - upperM))
+
+                let hi :: Int
+                    hi = h - 2 * max 0 (h - upperN)
+
+                let ks :: [Int]
+                    ks = [lo, lo + 2 .. hi]
+
+                forUntilJust ks $ \k -> do
+                  -- a = c[(k+1)%Z] if (k==-h or k!=h and c[(k-1)%Z]<c[(k+1)%Z]) else c[(k-1)%Z]+1
+                  initAVal <- do
+                    let part1 = k == -h
+
+                    let part2 = k /= h
+
+                    -- (k+1)%Z
+                    let kp1Ix = (k + 1) `modPortable` upperZ
+
+                    -- (k-1)%Z
+                    let km1Ix = (k - 1) `modPortable` upperZ
+                    if part1
+                      then MV.unsafeRead c kp1Ix
```

Comment:
What would happen if the index was out of bounds? Would it result in a garbage answer, or potentially a segfault? I see that it can't be out of bounds in this context, because it's taken mod `upperZ`, but I'm curious.

Reply:
Okay, that makes sense; it could segfault but it would most likely return garbage.